crates/guest-rust/src/lib.rs

```diff
@@ -681,23 +681,31 @@
 ///     // already has generated bindings for all WASI types and structures. In this
 ///     // situation the key `with` here can be used to use those types
 ///     // elsewhere rather than regenerating types.
+///     // If for example your world refers to some type and you want to use
+///     // your own custom implementation of that type then you can specify
+///     // that here as well. There is a requirement on the remapped (custom)
+///     // type to have the same internal structure and identical to what would
+///     // wit-bindgen generate (including alignment, etc.), since
+///     // lifting/lowering uses its fields directly.
 ///     //
 ///     // If, however, your world refers to interfaces for which you don't have
 ///     // already generated bindings then you can use the special `generate` value
 ///     // to have those bindings generated.
 ///     //
-///     // The `with` key only supports replacing types at the interface level
-///     // at this time.
+///     // The `with` key here works for interfaces and individual types.
 ///     //
-///     // When an interface is specified no bindings will be generated at
-///     // all. It's assumed bindings are fully generated somewhere else. This is an
-///     // indicator that any further references to types defined in these
-///     // interfaces should use the upstream paths specified here instead.
+///     // When an interface or type is specified here no bindings will be
+///     // generated at all. It's assumed bindings are fully generated
+///     // somewhere else. This is an indicator that any further references to types
+///     // defined in these interfaces should use the upstream paths specified
+///     // here instead.
 ///     //
 ///     // Any unused keys in this map are considered an error.
 ///     with: {
 ///         "wasi:io/poll": wasi::io::poll,
-///         "some:package/my-interface": generate,
+///         "some:package/my-interface1": types::MyInterface1Type,
```

Comment:
I was a little confused by the example on this line, since I'd expect a path on the right-hand side (since the left-hand side refers to an interface), rather than a type. Was this a typo, or am I misunderstanding how the macro distinguishes between types and interfaces?

Reply:
Nice! Glad it turned out that there was something more to it, since I'm sure it would've come up in the upstream review. I took another read through everything, and the documentation seems a lot clearer to me now. No notes on the other changes, but looking forward to seeing what upstream has to say!